docs/dev/design/new-search-api.rst

```diff
@@ -0,0 +1,298 @@
+ since ``:`` is already used to separate the key from the value.
+
+Including subprojects
+~~~~~~~~~~~~~~~~~~~~~
+
+Now that we are returning results only
+from the given projects, we need an easy way to
+include results from subprojects.
+Some ideas for implementing this feature are:
+
+``include-subprojects:true``
+  This doesn't make it clear from what
+  projects we are going to include subprojects from.
+  We could make it so it returns subprojects for all projects.
+  Users will probably use this with one project only.
+
+``subprojects:project/version`` (inclusive)
+  This allows to specify from what project exactly
+  we are going to return subprojects from,
+  and also include the version we are going to try to match.
+  This includes the parent project in the results.
+
+  As the ``project`` parameter, the version can be optional,
+  and defaults to the default version of the parent project.
+
+``subprojects:project/version`` (exclusive)
+  This is the same as the above,
+  but it doesn't include the parent project in the results.
+  If we want to include the results from the project, then
+  the query will be ``project:project/latest subprojects:project/latest``.
+  Is this useful?
+
+Cache
+-----
+
+Since the request could be attached to more than one project.
+We will return all the list of projects for the cache tags,
+this is ``project1, project1:version, project2, project2:version``.
+
+CORS
+----
+
+Since the request could be attached to more than one project.
+we can't make the decision if we should enable CORS or not on a given request from the middleware easily,
+so we won't allow cross site requests when using the new API for now.
+We would need to refactor our CORS code,
+so every view can decide if CORS should be allowed or not,
+for this case, cross site requests will be allowed only if all versions of the final search are public,
+another alternative could be to always allow cross site requests,
+but when a request is cross site, we only return results from public versions.
+
+Analytics
+---------
+
+We will record the same query for each project that was used in the final search.
+
+Response
+--------
+
+The response will be similar to the old one,
+but will include extra information about the search,
+like the projects, versions, and the query that were used in the final search.
+
+And the ``version``, ``project``, and ``project_alias`` attributes will
+now be objects.
+
+We could just re-use the old response too,
+since the only breaking changes would be the attributes now being objects,
+and we aren't adding any new information to those objects (yet).
+But also, re-using the current serializers shouldn't be a problem either.
+
+.. code-block:: json
+
+   {
+     "count": 1,
+     "next": null,
+     "previous": null,
+     "projects": [
+       {
+         "slug": "docs",
+         "versions": [
+           {
+             "slug": "latest"
+           }
+         ]
+       }
+     ],
```

Comment:
> Are we going to list here all the projects where the query was performed?

yes

> So, for project:docs project:dev project:test-builds this output will have 3 project objects?

yes

> How this output would be for project:docs subprojects:docs?

it will have the docs project and all the subprojects that were used in the query